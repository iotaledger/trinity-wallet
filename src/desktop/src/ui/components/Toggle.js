import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import css from './toggle.scss';

/**
 * Toggle component
 */
export default class Toggle extends React.PureComponent {
    static propTypes = {
        /** Checkbox state */
        checked: PropTypes.bool.isRequired,
        /** On state label */
        on: PropTypes.string.isRequired,
        /** Off state label */
        off: PropTypes.string.isRequired,
        /**Toggle state event function
         * @param {Bool} value - Current checkbox state
         */
        onChange: PropTypes.func.isRequired,
    };

    render() {
        const { off, on, checked, onChange } = this.props;

        return (
            <div className={classNames(css.toggle, checked ? css.on : null)} onClick={() => onChange(!checked)}>
                <span>{off}</span>
                <span>{on}</span>
            </div>
        );
    }
}
