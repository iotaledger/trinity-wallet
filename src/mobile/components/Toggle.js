import React, { PureComponent } from 'react';
import { View, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';
import { width } from '../utils/dimensions';
import { isAndroid } from '../utils/device';

const styles = StyleSheet.create({
    toggle: {
        borderWidth: 1.5,
        justifyContent: 'center',
    },
    toggleContainer: {
        alignItems: 'center',
        justifyContent: 'center',
    },
    toggleCircle: {},
});

class Toggle extends PureComponent {
    static propTypes = {
        /** Determines whether the toggle is active or not */
        active: PropTypes.bool.isRequired,
        /** @ignore */
        bodyColor: PropTypes.string.isRequired,
        /** @ignore */
        primaryColor: PropTypes.string.isRequired,
        scale: PropTypes.number,
        opacity: PropTypes.number,
    };

    static defaultProps = {
        scale: 1,
        opacity: 1,
    };

    render() {
        const { active, primaryColor, bodyColor, scale, opacity } = this.props;
        const size = width * scale;
        return (
            <View style={[styles.toggleContainer, { width: size / 12, height: size / 12 }]}>
                <View
                    style={[
                        styles.toggle,
                        {
                            alignItems: active ? 'flex-end' : 'flex-start',
                            borderColor: bodyColor,
                            width: size / 13,
                            height: size / 22,
                            borderRadius: size / 40,
                            paddingHorizontal: size / 300,
                            opacity: isAndroid ? opacity * 2 : 1,
                        },
                    ]}
                >
                    <View
                        style={[
                            styles.toggleCircle,
                            {
                                backgroundColor: primaryColor,
                                width: size / 33,
                                height: size / 33,
                                borderRadius: size / 33,
                            },
                        ]}
                    />
                </View>
            </View>
        );
    }
}

export default Toggle;
