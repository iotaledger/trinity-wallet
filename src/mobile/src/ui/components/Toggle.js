import React, { PureComponent } from 'react';
import { View, StyleSheet, Animated } from 'react-native';
import PropTypes from 'prop-types';
import { width } from 'libs/dimensions';
import { isAndroid } from 'libs/device';

const styles = StyleSheet.create({
    toggle: {
        borderWidth: 1.5,
        justifyContent: 'center',
    },
    toggleContainer: {
        alignItems: 'center',
        justifyContent: 'center',
    },
    toggleCircle: {},
});

class Toggle extends PureComponent {
    static propTypes = {
        /** Determines whether the toggle is active or not */
        active: PropTypes.bool.isRequired,
        /** @ignore */
        bodyColor: PropTypes.string.isRequired,
        /** @ignore */
        primaryColor: PropTypes.string.isRequired,
        /** Toggle scale multiplier */
        scale: PropTypes.number,
        /** Toggle opacity */
        opacity: PropTypes.number,
    };

    static defaultProps = {
        scale: 1,
        opacity: 1,
    };

    constructor(props) {
        super(props);
        this.state = {
            togglePosition: new Animated.Value(this.getTogglePosition(props.active)),
            toggleStyle: this.getToggleStyle(),
        };
    }

    componentWillReceiveProps(newProps) {
        if (this.props.active !== newProps.active) {
            Animated.timing(this.state.togglePosition, {
                toValue: this.getTogglePosition(!this.props.active),
                duration: 80,
            }).start(() => this.setState({ toggleStyle: this.getToggleStyle() }));
        }
    }

    /**
     * Gets toggle style depending on whether it is active
     *
     * @returns {object}
     */
    getToggleStyle() {
        const { active, scale, primaryColor, bodyColor } = this.props;
        if (active) {
            return { backgroundColor: primaryColor, borderWidth: 0 };
        }
        return { backgroundColor: 'transparent', borderWidth: scale * 1.5, borderColor: bodyColor };
    }

    /**
     * Gets toggle position depending on the given condition
     *
     * @param {bool} condition
     * @returns {object}
     */
    getTogglePosition(condition) {
        const size = width * this.props.scale;
        return condition ? size / 12 - size / 33 - 5 : 1.5;
    }

    render() {
        const { bodyColor, scale, opacity } = this.props;
        const size = width * scale;
        return (
            <View style={[styles.toggleContainer, { width: size / 12, height: size / 12 }]}>
                <View
                    style={[
                        styles.toggle,
                        {
                            borderColor: bodyColor,
                            width: size / 12,
                            height: size / 22,
                            borderRadius: size / 24,
                            opacity: isAndroid ? opacity * 2 : 1,
                        },
                    ]}
                >
                    <Animated.View
                        style={[
                            styles.toggleCircle,
                            {
                                width: size / 33,
                                height: size / 33,
                                borderRadius: size / 33,
                                position: 'absolute',
                                left: this.state.togglePosition,
                            },
                            this.state.toggleStyle,
                        ]}
                    />
                </View>
            </View>
        );
    }
}

export default Toggle;
