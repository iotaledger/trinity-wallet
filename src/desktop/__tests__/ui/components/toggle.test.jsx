import React from 'react';
import { shallow } from 'enzyme';

import Toggle from 'ui/components/Toggle';

const props = {
    checked: false,
    on: 'On',
    off: 'Off',
    onChange: jest.fn(),
};

describe('Toggle component', () => {
    test('Render the component', () => {
        const wrapper = shallow(<Toggle {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('Off state', () => {
        const wrapper = shallow(<Toggle {...props} />);

        expect(wrapper.find('div div').hasClass('on')).toBeFalsy();
    });

    test('On state', () => {
        const mockProps = Object.assign({}, props, { checked: true });
        const wrapper = shallow(<Toggle {...mockProps} />);

        expect(wrapper.find('div div').hasClass('on')).toBeTruthy();
    });

    test('Toggle event', () => {
        const wrapper = shallow(<Toggle {...props} />);

        wrapper.simulate('click');
        expect(props.onChange).toHaveBeenCalledTimes(1);
    });
});
