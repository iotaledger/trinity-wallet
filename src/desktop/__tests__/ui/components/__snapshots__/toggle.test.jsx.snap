// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle component Render the component 1`] = `
<div
  className="toggle"
  onClick={[Function]}
>
  <div
    className="active"
  >
    <span>
      Off
    </span>
    <span>
      On
    </span>
  </div>
</div>
`;
